### ISSUE TITLE
Crash when selecting the analyzer

### ISSUE THREAD

**Issue by ecrucru** [pychess/pychess#1484]
Hello,

**Steps to reproduce**

- Install 2 engines of your choice. For me :
    - Windows: pyChess 0.99, Stockfish 8 (stockfishchess.org) and Komodo 9.0.3 (komodochess.com)
    - Linux: latest pyChess git, Stockfish (official-stockfish/Stockfish) and Stockfish Variants (ddugovic/stockfish)
- Run pyChess
- Define Stockfish as the analyzer for both sides in the options
- Start a new game : human vs. human, normal chess
- Activate the hint and spy arrows from the menu bar
- Play few moves to engage the game
- Menu Action > Resign : the game is finished
- With the widget score, take another position to rewind the elapsed game
- The engines start to analyze the chosen position
- Go to the options and select Komodo as the hint analyzer (change several times if needed)

**Windows**

pyChess crashes and the data are lost ! Microsoft invites me to send the report :

![image](https://user-images.githubusercontent.com/24614488/31865530-236dd600-b770-11e7-8f4f-c54c53055d7a.png)

The saved log shows :

```
[...the engines are playing a lot...]
unknown DEBUG: (pychess.exe:2384): GdkPixbuf-[1;32mDEBUG[0m: gdk_pixbuf_from_pixdata() called on:
unknown DEBUG: (pychess.exe:2384): GdkPixbuf-[1;32mDEBUG[0m: Encoding raw
unknown DEBUG: (pychess.exe:2384): GdkPixbuf-[1;32mDEBUG[0m: Dimensions: 16 x 16
unknown DEBUG: (pychess.exe:2384): GdkPixbuf-[1;32mDEBUG[0m: Rowstride: 64, Length: 1048
unknown DEBUG: (pychess.exe:2384): GdkPixbuf-[1;32mDEBUG[0m: Copy pixels == false
unknown WARNING: Executing <TimerHandle when=1015658.525 glib_update(<GLib.MainCon...t 0x0149D2A0)>, <ProactorEven...se debug=True>) at pychess:223 created at pychess:226> took 0.157 seconds
('Stockfish8.exe', '-') DEBUG: info depth 6 seldepth 6 multipv 3 score cp 34 nodes 2839 nps 315444 tbhits 0 time 9 pv c7c5 g1e2 d7d5 d2d4 g8f6 f1g2
('Komodo902.exe',  '-') DEBUG: info multipv 2 depth 5 time 3140 nodes 8075 score cp 11 nps 2571 tbhits 0 pv b1c3 g8f6 d2d4 d7d5 c1d2
unknown WARNING: Executing <TimerHandle when=1015658.76 glib_update(<GLib.MainCon...t 0x0149D2A0)>, <ProactorEven...se debug=True>) at pychess:223 created at pychess:226> took 0.156 seconds
('Stockfish8.exe', '-') DEBUG: info depth 7 seldepth 7 multipv 1 score cp 40 nodes 5232 nps 348800 tbhits 0 time 15 pv g8f6 b1c3 d7d5 d2d4 c7c5 g1f3 c8d7
('Komodo902.exe',  '-') DEBUG: info multipv 3 depth 4 time 3140 nodes 8075 score cp 1 nps 2571 tbhits 0 pv g1f3 d7d5 d2d4 b8c6
unknown WARNING: Executing <TimerHandle when=1015658.9940000001 glib_update(<GLib.MainCon...t 0x0149D2A0)>, <ProactorEven...se debug=True>) at pychess:223 created at pychess:226> took 0.110 seconds
('Stockfish8.exe', '-') DEBUG: info depth 7 seldepth 7 multipv 2 score cp 40 nodes 5232 nps 348800 tbhits 0 time 15 pv d7d5 d2d4 g8f6 b1c3 c7c5 g1f3 c8d7
('Komodo902.exe',  '-') DEBUG: info multipv 1 depth 5 time 3140 nodes 8433 score cp 11 nps 2685 tbhits 0 pv d2d4 g8f6 b1c3 d7d5 c1d2
('Stockfish8.exe', '-') DEBUG: info depth 7 seldepth 7 multipv 3 score cp 28 nodes 5232 nps 348800 tbhits 0 time 15 pv c7c5 g1f3 d7d5 f1g2 b8c6 d2d4 g8f6
('Komodo902.exe',  '-') DEBUG: info multipv 2 depth 5 time 3140 nodes 8433 score cp 11 nps 2685 tbhits 0 pv b1c3 g8f6 d2d4 d7d5 c1d2
('Stockfish8.exe', '-') DEBUG: info depth 8 seldepth 10 multipv 1 score cp 22 nodes 8566 nps 389363 tbhits 0 time 22 pv g8f6 b1c3 d7d5 d2d4 f8e7 g1e2 e8g8 f1g2
('Komodo902.exe',  '-') DEBUG: info multipv 3 depth 5 time 3140 nodes 8433 score cp 3 nps 2685 tbhits 0 pv g1f3 g8f6 b1c3 b8c6 d2d4
('Stockfish8.exe', '-') DEBUG: info depth 8 seldepth 10 multipv 2 score cp 22 nodes 8566 nps 389363 tbhits 0 time 22 pv d7d5 d2d4 g8f6 b1c3 f8e7 g1e2 e8g8 f1g2
('Komodo902.exe',  '-') DEBUG: info multipv 1 depth 5 time 3140 nodes 8445 score cp 11 nps 2688 tbhits 0 pv d2d4 g8f6 b1c3 d7d5 c1d2
('Stockfish8.exe', '-') DEBUG: info depth 8 seldepth 10 multipv 3 score cp 19 nodes 8566 nps 389363 tbhits 0 time 22 pv c7c5 b1c3 d7d5 d2d4 g8f6 g1f3 c8d7 h2h3
('Komodo902.exe',  '-') DEBUG: info multipv 2 depth 5 time 3140 nodes 8445 score cp 11 nps 2688 tbhits 0 pv b1c3 g8f6 d2d4 d7d5 c1d2
('Stockfish8.exe', '-') DEBUG: info depth 9 seldepth 10 multipv 1 score cp 27 nodes 13239 nps 401181 tbhits 0 time 33 pv g8f6 d2d4 d7d5 b1c3 c7c5 g1e2 c5d4 e3d4 b8c6
('Komodo902.exe',  '-') DEBUG: info multipv 3 depth 5 time 3140 nodes 8445 score cp 3 nps 2688 tbhits 0 pv g1f3 g8f6 b1c3 b8c6 d2d4
('Stockfish8.exe', '-') DEBUG: info depth 9 seldepth 10 multipv 2 score cp 22 nodes 13239 nps 401181 tbhits 0 time 33 pv d7d5 d2d4 g8f6 b1c3 f8e7 g1e2 e8g8 f1g2 b8c6
('Komodo902.exe',  '-') DEBUG: info multipv 1 depth 5 time 3140 nodes 8462 score cp 11 nps 2694 tbhits 0 pv d2d4 g8f6 b1c3 d7d5 c1d2
unknown ERROR: C:\pychess\library.zip\gi\overrides\GObject.py:553: Warning: gsignal.c:2641: instance '016EDE10' has no handler with id '35214'
```

The same error also appears when I close the main window of pyChess while it analyzes a PGN game that I have opened. There is no need here to change the analyzer in the settings to get the crash.

I use these settings :

![image](https://user-images.githubusercontent.com/24614488/31865497-1423b8ae-b770-11e7-9192-3d642e240ea9.png)

![image](https://user-images.githubusercontent.com/24614488/31865525-1d720032-b770-11e7-826d-2a034512b88c.png)

Note : the crash only occurs when the analyzer is thinking, not when it has finished ! That's why you must increase the number of seconds of analysis to maximize the probability of the bug.

**Linux**

The application is not dead, but it displays this error message in the console :

```
/usr/lib/python3/dist-packages/gi/overrides/GObject.py:536: Warning: /build/glib2.0-EvFudu/glib2.0-2.42.1/./gobject/gsignal.c:2579: instance '0xa044a50' has no handler with id '24528'
  return func(*args, **kwargs)
```

In other words : soft error under Linux, hard error under Windows.

Regards

**Comment by gbtami** [comment]
Can you provide what exact version/binarie of engines you used?
Is this reprodicible with current git also?
Add --log-debug command line arg to pychess starting icon and attach the .log please!

**Comment by ecrucru** [comment]
*(Edit: merged into the first comment)*

**Comment by gbtami** [comment]
Unfortunately I was not able to reproduce this yet. 
Can you paste a screen shot of your
preferences - general
preferences - hints
and the whole pychess when it crashes, please?

Some more question. Can you reproduce this with other engines, like glaurung, toga2, etc. also?
What Linux distrib and DE are you using?

**Comment by ecrucru** [comment]
*(Edit: merged into the first comment)*

**Comment by gbtami** [comment]
Can you explain a bit more detailed what exactly happens when pychess crashes on Linux? Is it frozen or closed or segfaults or ... ? What can you see in the console?

What 4. "Abandon the game" is meaning? Give up/draw/close/...?

You said that "It occurs also in the latest git under Linux, but the main process is not terminated."
Is this means that pychess window is still open? If yes, can you create a screenshot on Linux after pychess crashes please?

**Comment by ecrucru** [comment]
I have merged the comments into the first post.
Tell me if you need more details again.
Thanks

**Comment by gbtami** [comment]
You said that on Linux "The application is not dead, but it displays this error message in the console: ...."
That message is a Warning only. Am I understand correctly that on Linux after changing the analyzer pychess works OK, meaning it's showing PV lines from new analyzer and you can continue to work with it, select another position etc? All in all, is this a Windows only crash?

**Comment by gbtami** [comment]
Another question. As you can see at https://github.com/pychess/pychess/blob/master/lib/pychess/System/SubProcess.py#L134 terminate() logs itself in a try except block if you started pychess with --log-debug
Can you search in .log file for "SubProcess.terminate():" strings please? Is it followed by "ProcessLookupError"  on Windows?

**Comment by ecrucru** [comment]
Hello,

My issue is specific to Windows. It doesn't crash on Linux.

I followed your instructions and I finally arrived at this observation :

- `pychess.exe` = no bug
- `pychess.exe --log-viewer` = **crash**
- `pychess.exe --log-debug` = no bug
- `pychess.exe --log-viewer --log-debug` = **crash**

My default shortcut was influenced by your comment on #1440. I didn't figure out that it could have any influence of that kind.

I really hope that you will be able to reproduce the case now.

Thanks

**Comment by gbtami** [comment]
Thx! Adding --log-viewer I was able to reproduce the crash finally.

**Comment by gbtami** [comment]
Seems this bug appears in frozen pychess (created with cx_freezee) .msi only. I'v just committed a fix, and it seems OK here. I will upload a .msi somewhere to let you test it also.
https://drive.google.com/open?id=0B0rQUiKQl0Qvc3R3cXAxU3lTWGM
@ecrucru test it please!

**Comment by ecrucru** [comment]
Hello,

The use of psutil above 3.4.2 drops the support for Windows XP and pyChess will not run anymore on it (2000 and 2003 as well). I don't have any other Windows-based computer for which I am *allowed* to install pyChess to test what you changed. Sorry for that.

In any case, it is nice if you reproduced the case.

Regards

**Comment by gbtami** [comment]
An updated one with psutil 3.4.2 https://drive.google.com/open?id=0B0rQUiKQl0QvaTNkZi1ZVlk2bm8

**Comment by ecrucru** [comment]
Everything is alright. Good job again, thank you.
psutil is effective because the analyzers have a lower priority.

**Comment by gbtami** [comment]
Thx for testing it!
